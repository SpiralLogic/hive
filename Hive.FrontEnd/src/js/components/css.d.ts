/* eslint-disable no-unused-vars */
// noinspection ES6UnusedImports
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import * as React from 'react';

declare module 'react' {
    interface CSSProperties {
        '--color'?: string;
        '--hex-offset'?: number;
        '--hex-size'?: string;
    }
}
