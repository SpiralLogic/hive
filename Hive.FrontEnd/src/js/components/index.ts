export * from './Cell';
export * from './GameArea';
export * from './Hextille';
export * from './PlayerTiles';
export * from './PlayerList';
export * from './App';
export * from './Row';
export * from './Tile';
