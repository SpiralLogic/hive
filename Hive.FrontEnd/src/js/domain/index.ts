export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './hex-coordinates';
