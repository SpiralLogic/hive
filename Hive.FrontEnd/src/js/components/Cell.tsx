import * as React from 'react';
import { useDrop } from 'react-dnd';
import { Hexagon, HexCoordinates } from '../domain';
import { Tile, TILE_TYPE } from './Tile';

type CellProps = Hexagon

const areEqual = (a: HexCoordinates, b: HexCoordinates) => a && b && a.q === b.q && a.r === b.r;
const isValidMove = (move: HexCoordinates, validMoves: HexCoordinates[]) => validMoves.some(dest => areEqual(move, dest));
const coordinateToString = ({ q, r }: HexCoordinates) => `${q}-${r}`;

export const Cell: React.FunctionComponent<CellProps> = ({
    tiles,
    coordinates,
}) => {
    const [{ isOver, canDrop }, drop] = useDrop({
        accept: TILE_TYPE,
        drop: () => coordinates,
        canDrop: (item, monitor) => isValidMove(coordinates, monitor.getItem().availableMoves),
        collect: monitor => ({
            isOver: monitor.isOver(),
            canDrop: monitor.canDrop(),
        }),
    });

    const backGroundColor = () => {
        if (canDrop && isOver) return ' valid-cell active';
        if (isOver) return ' invalid-cell';
        if (canDrop) return ' valid-cell';
        return '';
    };
    const id = coordinateToString(coordinates);
    return (
        <div className={'hex cell' + backGroundColor()} ref={drop}>
            {tiles.length > 0 && <Tile key={'tile' + id} {...tiles[0]} />}
        </div>
    );
};

Cell.displayName = 'Cell';