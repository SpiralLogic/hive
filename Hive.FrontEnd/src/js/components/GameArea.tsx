import { h } from 'preact';
import { useState } from 'preact/compat';
import { useEffect } from 'preact/hooks';
import { GameState } from '../domain';
import { CellDropEvent, useCellDropEmitter } from '../emitters/cell-drop-emitter';
import Engine from '../game-engine';
import { handleDragOver } from '../handlers';
import Hextille from './Hextille';
import PlayerList from './PlayerList';

const cellDropEmitter = useCellDropEmitter();

const GameArea = () => {
    const [gameState, setGameState] = useState<GameState | undefined>(undefined);

    const attributes = {
        ondragover: handleDragOver,
        className: 'hive',
        style: { '--hex-size': '50px' },
    };

    useEffect(() => {
        const fetch = (async () => {
            setGameState(await Engine.newGame());
        });
        fetch();
    }, []);

    useEffect(() => {
        const cellDropListener = async (e: CellDropEvent) => setGameState(await Engine.moveTile(e));

        cellDropEmitter.add(cellDropListener);
        return () => cellDropEmitter.remove(cellDropListener);
    }, [gameState]);

    if (!gameState) {
        return <h1>loading !</h1>;
    }

    return (
        <div {...attributes}>
            <PlayerList players={gameState.players}/>
            <Hextille cells={gameState.cells}/>
        </div>
    );
};

GameArea.displayName = 'GameArea';
export default GameArea;
