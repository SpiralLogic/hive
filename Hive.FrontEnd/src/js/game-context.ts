import { createContext, useContext, useEffect, useMemo, useState } from 'preact/compat';
import { GameState, Hexagon, Move, MoveTile, Player } from './domain';
import { Engine } from './game-engine';

interface GameContext {
    players: Player[];
    hexagons: Hexagon[];
    moveTile: MoveTile;
}

export const useNewHiveContext = (): [boolean, GameContext?] => {
    const [loading, setLoading] = useState(true);
    const [gameState, setGameState] = useState<GameState>(undefined!);

    useEffect(() => {
        Engine.initialState()
            .then((state) => {
                setGameState(state);
                return setLoading(false);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, []);

    const moveTile = useMemo(() => (move: Move) => {
        Engine.moveTile(move)
            .then((state) => {
                return setGameState(state);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, [setGameState]);

    if (loading) return [loading];

    return [loading, { ...gameState, moveTile }];
};

export const HiveContext = createContext<GameContext>({
    hexagons: [],
    players: [],
    moveTile: () => undefined,
});

export const useHiveContext = () => useContext(HiveContext);
// @ts-ignore
HiveContext.displayName = 'Hive Context';
