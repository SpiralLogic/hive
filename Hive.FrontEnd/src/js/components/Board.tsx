import * as React from 'react';
import { coordinateAsId, ICell } from '../domain';
import { Context } from '../GameContext';
import { Cell } from './Cell';

const getHextilleWidthDimensions = (cells: ICell[]) => {
    const [min, max] = cells
        .reduce(([min, max], c) => [Math.min(min, c.coordinates.q), Math.max(max, c.coordinates.q)], [0, 0]);

    return [min, max - min + 1];
};
export const Board: React.FunctionComponent = () => {
    const { allCells } = React.useContext(Context);
    const sortedCells = allCells.sort((c1, c2) => c1.coordinates.r - c2.coordinates.r || c1.coordinates.q - c2.coordinates.q);
    const firstRow = sortedCells[0].coordinates.r;
    const height = sortedCells[sortedCells.length - 1].coordinates.r - firstRow + 1;
    const [minQ, width] = getHextilleWidthDimensions(sortedCells);
    const columnShift = { '--hex-offset': (firstRow % 2) ? -1 : 1 };

    const rowCells = sortedCells.reduce((rows, cell) => {
        rows[cell.coordinates.r - firstRow][cell.coordinates.q - minQ] = cell;
        return rows;
    }, Array.from(Array(height), () => new Array(width).fill(false)));

    return (
        <div className="board" style={columnShift}>
            {rowCells.map((cells, r) => (
                <div key={r} className="hex-row">
                    {cells.map((cell, q) =>
                        cell ? <Cell key={coordinateAsId(cell.coordinates)} {...cell}/> :
                            <div key={q + ' ' + r} className={'hidden'}/>)}
                </div>
            ))}
        </div>
    );
};