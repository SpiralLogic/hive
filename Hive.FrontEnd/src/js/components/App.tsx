import {GameArea} from './GameArea';
import { h } from 'preact';

export default () => {
    return <GameArea/>;
};
