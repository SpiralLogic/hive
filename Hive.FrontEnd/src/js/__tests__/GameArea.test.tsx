import { create } from 'react-test-renderer';
import * as React from 'react';
import { Hexagon,  Player } from '../domain';
import * as TestUtils from 'react-dom/test-utils';
import { render } from 'react-dom';
import { GameArea } from '../components';
import * as CTX from '../game-context';
import { HiveContext } from '../game-context';
let props: { loading: boolean; players: Player[]; hexagons: Hexagon[] };

beforeEach(() => {
    const hexagon = { coordinates: { q: 1, r: 1 }, tiles: [] };
    const player: Player = {
        id: 1,
        name: 'adsf',
        availableTiles: [{ id: 1, playerId: 1, availableMoves: [], content: 'ant' }],
    };

    const gameState = { hexagons: [hexagon], players: [player] };

    props = { players: [player], hexagons: [hexagon], loading: false };
    jest.spyOn(CTX, 'useHiveContext').mockImplementation(() => ({ moveTile: jest.fn(), ...gameState }));
});

test('Game area shows initial loading', () => {
    props.loading = true;
    const component = create(<GameArea loading={true} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});

test('default on drop is prevented', async () => {
    await TestUtils.act(async () => {
        const container = document.createElement('div');
        document.body.appendChild(container);

        render(
            <HiveContext.Provider value={CTX.useHiveContext()}>
                <GameArea loading={false} />
            </HiveContext.Provider>,
            container,
        );
    });

    const gameArea = document.querySelectorAll('.hive')[0];
    const preventDefault = jest.fn();
    TestUtils.Simulate.dragOver(gameArea, { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
});

test('Game area shows board after fetch', async () => {
    const component = create(<GameArea loading={false} />);

    expect(component.toJSON()).toMatchSnapshot();
});
