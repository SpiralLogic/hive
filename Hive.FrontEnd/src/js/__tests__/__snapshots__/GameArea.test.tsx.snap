// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows loading 1`] = `
<h1>
  loading !
</h1>
`;
